compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -903,67 +903,64 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
-                    .chain(
-                        data_a
-                            .projection_bounds()
-                            .map(|b| b.map_bound(ty::ExistentialPredicate::Projection)),
-                    )
-                    .chain(
-                        data_b
-                            .auto_traits()
-                            .map(ty::ExistentialPredicate::AutoTrait)
-                            .map(ty::Binder::dummy),
-                    );
-                let existential_predicates = tcx.mk_poly_existential_predicates_from_iter(iter);
-                let source_trait = Ty::new_dynamic(tcx, existential_predicates, r_b, repr_a);
+        let ty::Dynamic(data_a, r_a, ty::Dyn) = *source.kind() else {
+            bug!();
+        };
+        let ty::Dynamic(data_b, r_b, ty::Dyn) = *target.kind() else {
+            bug!();
+        };
 
-                // Require that the traits involved in this upcast are **equal**;
-                // only the **lifetime bound** is changed.
-                let InferOk { obligations, .. } = self
-                    .infcx
-                    .at(&obligation.cause, obligation.param_env)
-                    .sup(DefineOpaqueTypes::No, target, source_trait)
-                    .map_err(|_| Unimplemented)?;
-                nested.extend(obligations);
+        let principal_a = data_a.principal().unwrap();
+        let source_trait_ref = principal_a.with_self_ty(tcx, source);
+        // The supertrait of the source's principal that we'd like to match against our target.
+        let upcast_trait_ref = util::supertraits(tcx, source_trait_ref).nth(idx).unwrap();
 
-                let outlives = ty::OutlivesPredicate(r_a, r_b);
-                nested.push(Obligation::with_depth(
-                    tcx,
-                    obligation.cause.clone(),
-                    obligation.recursion_depth + 1,
-                    obligation.param_env,
-                    obligation.predicate.rebind(outlives),
-                ));
-            }
-            _ => bug!(),
-        };
+        // The trait ref that we're casting to.
+        let target_trait_ref = data_b.principal().unwrap().with_self_ty(tcx, source);
+
+        // Relate the principals, to make sure that the supertrait we found in
+        // the target actually satisfies the principal of the target.
+        // FIXME: This should be very easy to support "upcasting" from
+        // `dyn Trait + Send` to `dyn Send` with an if-let.
+        let InferOk { obligations: mut nested, .. } = self
+            .infcx
+            .at(&obligation.cause, obligation.param_env)
+            .sup(DefineOpaqueTypes::No, target_trait_ref, upcast_trait_ref)
+            .map_err(|_| Unimplemented)?;
+
+        // Secondly, make sure that all of the non-principal bounds of the
+        // target (associated types and auto traits) are satisfied by the source.
+        nested.extend(data_b.projection_bounds().map(|bound| {
```

Comment:
same here, feel like upcasting is safer to impl by iterating over all existential predicates of `b`